sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -71,128 +82,111 @@ fn parse_hex_str(hex_str: &str) -> Result<Vec<u8>> {
-        std::array::IntoIter::new(u32::try_from(image.len())?.to_le_bytes()),
-    )?;
+fn update_image_manifest(image: &mut Image, key: &PrivateKey) -> Result<()> {
+    image.manifest.identifier = 0x4552544f;
+    image.manifest.reserved0 = 0;
+    image.manifest.image_length = u32::try_from(image.bytes().len())?;
+    image.manifest.image_version = 0;
+    image.manifest.image_timestamp = 0;
+    const EXPONENT_SIZE: usize = size_of::<u32>();
+    let exponent_be = key.public_exponent_be();
+    ensure!(
+        exponent_be.len() <= EXPONENT_SIZE,
+        "Public exponent ({} bytes) is longer than {} bytes.",
+        exponent_be.len(),
+        EXPONENT_SIZE,
+    );
+    image.manifest.exponent = u32::from_be_bytes(
+        iter::repeat(0)
+            .take(EXPONENT_SIZE - exponent_be.len())
+            .chain(exponent_be.iter().cloned())
+            .collect::<Vec<_>>()
+            .as_slice()
+            .try_into()?,
+    );
+    image.manifest.reserved1 = 0;
+    image.manifest.usage_constraints = [0; 8usize];
+    image.manifest.lockdown_info = [0; 4usize];
+    let mut modulus = key.public_modulus_be();
+    modulus.reverse();
+    image
+        .manifest
+        .modulus
+        .as_bytes_mut()
+        .copy_from_slice(&modulus);
+    image.manifest.extension0_offset = 0;
+    image.manifest.extension0_checksum = 0;
+    image.manifest.extension1_offset = 0;
+    image.manifest.extension1_checksum = 0;
+    image.manifest.extension2_offset = 0;
+    image.manifest.extension2_checksum = 0;
+    image.manifest.extension3_offset = 0;
+    image.manifest.extension3_checksum = 0;
 
     Ok(())
 }
 
 /// Calculates the signature for the signed portion of an image.
 fn calculate_image_signature(
     image: &Image,
-    key: impl AsRef<Path>,
+    private_key: &PrivateKey,
 ) -> Result<ImageSignature> {
-    let key = fs::read(&key).with_context(|| {
-        format!("Failed to read key from `{}`.", key.as_ref().display())
-    })?;
-    let key =
-        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
-    let sig = ImageSignature::sign(&key, image.signed_bytes())
-        .context("Failed to sign the image.")?;
-    Ok(sig)
+    ImageSignature::sign(&private_key, &image.signed_bytes())
+        .context("Failed to calculate image signature.")
 }
 
 /// Updates the signature of an image.
-fn update_image_signature(
-    image: &mut Image,
-    signature: ImageSignature,
-) -> Result<()> {
-    image.set_manifest_field(
-        &manifest::ROM_EXT_IMAGE_SIGNATURE,
-        signature.bytes().iter().rev().cloned(),
-    )?;
-    Ok(())
+fn update_image_signature(image: &mut Image, sig: ImageSignature) {
+    image
+        .manifest
+        .signature
+        .data
+        .as_bytes_mut()
+        .copy_from_slice(
+            sig.bytes()
+                .iter()
+                .rev()
+                .cloned()
+                .collect::<Vec<_>>()
+                .as_slice(),
+        );
```

Comment:
```
let dest = image....as_bytes_mut().iter_mut();
let src = sig.bytes().iter().rev().copied();
for (d, s) in Iterator::zip(dest, src) {
  *d = s;
}
```

You can throw in a length check if you care enough. =)

Reply:
Thanks, this doesn't need to `collect`. Done.